src/components/DpAutocomplete/DpAutocomplete.vue

```diff
@@ -1,114 +1,396 @@
+          range.collapse(true)
+          sel?.removeAllRanges()
+          sel?.addRange(range)
+        }
       }
     }
+    el.focus()
+  })
+}
+
+function getTextContent(el = input.value) {
+  return el?.textContent || ''
+}
+
+function runSpecialKeys(e: KeyboardEvent) {
+  const key = e.key
+
+  if (key === 'Tab' && completion.value) {
+    e.preventDefault()
+    currentQuery.value += completion.value
+    setTextContent(currentQuery.value)
+    focusInput()
+    emit('update:modelValue', currentQuery.value)
+  } else if (key === 'ArrowDown' || key === 'Down') {
+    e.preventDefault()
+    if (listPosition.value < props.options.length - 1) {
+      listPosition.value += 1
+    }
+  } else if (key === 'ArrowUp' || key === 'Up') {
+    e.preventDefault()
+    if (listPosition.value > -1) {
+      listPosition.value -= 1
+    }
+  } else if (key === 'Enter') {
+    e.preventDefault()
+    if (listPosition.value > -1 && props.options[listPosition.value]) {
+      selectCurrentOption(props.options[listPosition.value] as Record<string, unknown>)
+    } else {
+      triggerSearch()
+    }
+  } else if (key === 'Escape') {
+    e.preventDefault()
+    isOpenList.value = false
+  }
+}
+
+function selectCurrentOption (option: Record<string, unknown>) {
+  if (!option) return
+
+  currentQuery.value = option[props.label] as string
+  setTextContent(currentQuery.value)
+  listPosition.value = -1
+
+  emit('selected', option)
+  emit('update:modelValue', currentQuery.value)
+
+  isOpenList.value = false
+  focusInput()
+}
+
+function setTextContent (value: string, el = input.value) {
+  if (el) {
+    el.textContent = value
+  }
+}
+
+function triggerSearch () {
+  emit('searched', currentQuery.value)
+  isOpenList.value = false
+}
+
+function updateValue () {
+  const content = getTextContent()
+
+  if (content !== currentQuery.value) {
+    currentQuery.value = content
+    isOpenList.value = true
+    emit('update:modelValue', currentQuery.value)
+
+    if (content && content.length >= props.minChars) {
+      fetchOptions(content)
+    }
   }
 }
+
+async function fetchOptions(searchString: string) {
+  if (!searchString || searchString.length < props.minChars) {
+    return
+  }
+
+  isLoading.value = true
+
+  try {
+    const route = props.routeGenerator(searchString)
+    const response = await dpApi({
+      method: 'GET',
+      url: route,
+      params: {},
+      headers: {},
+      data: {}
+    })
+
+    // Only emit results that match the current search -> prevents race conditions
+    if (currentQuery.value === searchString) {
+      emit('search-changed', response)
+    }
+  } catch (e) {
+    console.error('Error fetching autocomplete options:', e)
+  } finally {
+    isLoading.value = false
+  }
+}
+
+// Setup mutation observer to track contenteditable changes
+onMounted(() => {
+  if (!input.value) {
+    return
+  }
+
+  const observer = new MutationObserver((mutations) => {
```

Comment:
should the MutationObserver be cleaned up in the onUnmounted hook?

Reply:
good point. Thanks for spotting
738d4666e9f0683579e3cb5d79d897f8bb55baae